decode.go

```diff
@@ -214,12 +214,23 @@ func (d *Decoder) DecodeValue() (interface{}, error) {
 	}
 }
 
+/////////////////////////////////////////
+// typeRefs
+/////////////////////////////////////////
+type TypeRefs struct {
+	typeRefs []reflect.Type
+	records  map[string]bool // record if existing for type
+}
+
 // appendTypeRefs add list or map type ref
-func (d *Decoder) appendTypeRefs(p reflect.Type) {
-	for i := 0; i < len(d.typeRefs); i++ {
-		if d.typeRefs[i] == p {
-			return
-		}
+func (t *TypeRefs) appendTypeRefs(name string, p reflect.Type) {
```

Comment:
concurrent read/write map issue may occur, better to change to `sync.Map`

Reply:
In a decoding, `Decoder` only have a `*TypeRefs`. `concurrent read/write map` will not be occured. |
And if it occur, the `typeRefs []reflect.Type` will also is unsafe.